src/xdist/dsession.py

```diff
@@ -182,7 +182,7 @@ def worker_workerfinished(self, node):
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
-        if node in self.sched.nodes:
+        if not self.shouldstop and node in self.sched.nodes:
```

Comment:
Without this gating, the following error would be occur using the example code in #1024:

```
INTERNALERROR> AssertionError: ('test_parallel.py::test_2', <WorkerController gw0>)
```

Reply:
I assume pytest-xdist interprets the worker existing early as it crashing? Is gw0 the worker that reached the local maxfail, or the other one?

Reply:
Yeah that's correct, `gw0` is the worker that is reaching the local maxfail in the example shown in my comment.

Reply:
Can you explain what is going on here then -- why is `node` still in `self.sched.nodes` even though it exited cleanly and not crashed?

Reply:
Just a reminder, I never really worked on the pytest-xdist code so I'm mostly trying to interpret things just like you :)

> My interpretation of self.sched.nodes is that it is a fixed-list, it doesn't get updated/reduced when workers stop running.

Right.

>  If the worker completed sucessfully, crashitem would be None.

This is true.

> In our case it did not complete succesfully - it aborted early - so crashitem is not None.

I think it *should* complete successfully, where I define "complete successfully" as "did not crash" (as opposed to "had any test failures" or such).

What is the worker <-> coordinator communication that needs to happen for the `shouldstop` scenario to not be considered a crash?

Reply:
> Just a reminder, I never really worked on the pytest-xdist code so I'm mostly trying to interpret things just like you :)

Understood! Sorry to keep tagging you, just not sure who else to get feedback from. Does this project have active maintainers?

> What is the worker <-> coordinator communication that needs to happen for the shouldstop scenario to not be considered a crash?

Sadly I don't really know. I see that, in `remote.py`, the worker calls `self.sendevent` in a number of scenarios - for example, `runtest_protocol_complete`, testreport`, `internal_error`, etc. I tried inserting some of these `sendevent` calls into the `pytest_runtestloop` to see if that leads to more graceful shutdown. Unforatunely nothing seems to work. I think I'm lacking sufficient knowledge to go any further.